src/main/kotlin/io/libp2p/core/multiformats/Protocol.kt

```diff
@@ -0,0 +1,157 @@
+package io.libp2p.core.multiformats
+
+import io.ipfs.cid.Cid
+import io.ipfs.multiaddr.Base32
+import io.libp2p.core.types.readUvarint
+import io.libp2p.core.types.toByteArray
+import io.libp2p.core.types.toByteBuf
+import io.libp2p.core.types.writeUvarint
+import io.netty.buffer.ByteBuf
+import java.net.Inet4Address
+import java.net.Inet6Address
+import java.nio.charset.StandardCharsets
+import io.netty.buffer.Unpooled.buffer as byteBuf
+
+
+private const val LENGTH_PREFIXED_VAR_SIZE = -1
+
+/**
+ * Partially translated from https://github.com/multiformats/java-multiaddr
+ */
+enum class Protocol(val code: Int, val size: Int, val typeName: String) {
+
+    IP4(4, 32, "ip4"),
+    TCP(6, 16, "tcp"),
+    UDP(273, 16, "udp"),
+    DCCP(33, 16, "dccp"),
+    IP6(41, 128, "ip6"),
+    IP6ZONE(42, LENGTH_PREFIXED_VAR_SIZE, "ip6zone"),
+    DNS4(54, LENGTH_PREFIXED_VAR_SIZE, "dns4"),
+    DNS6(55, LENGTH_PREFIXED_VAR_SIZE, "dns6"),
+    DNSADDR(56, LENGTH_PREFIXED_VAR_SIZE, "dnsaddr"),
+    SCTP(132, 16, "sctp"),
+    UTP(301, 0, "utp"),
+    UDT(302, 0, "udt"),
+    UNIX(400, LENGTH_PREFIXED_VAR_SIZE, "unix") {
+        override fun isPath() = true
+    },
+    IPFS(421, LENGTH_PREFIXED_VAR_SIZE, "ipfs"),
+    P2P(421, LENGTH_PREFIXED_VAR_SIZE, "p2p"),
+    HTTPS(443, 0, "https"),
+    ONION(444, 96, "onion"),
+    QUIC(460, 0, "quic"),
+    WS(477, 0, "ws"),
+    P2PCIRCUIT(290, 0, "p2p-circuit"),
+    HTTP(480, 0, "http");
+
+    val encoded: ByteArray = encode(code)
+
+    private fun encode(type: Int): ByteArray =
+        byteBuf(4).writeUvarint(type).toByteArray()
+
+    open fun isPath() = false
+
+    fun addressToBytes(addr: String): ByteArray =
```

Comment:
Eventually I'd like to get to a point where we use some form of transcoder functions + generics. So each protocol would be associated with a `data class` that encapsulates its fields/values, strongly typed. And we'd register the protocols using reified generics (super nice in Kotlin, I believe they translate to `Class<?>` parameters in plain Java). Then you can do something like:

```
val ma : Multiaddr = foo
val ipv4 : IPv4 = ma.first<IPv4>()
println(ipv4.address)
```

But let's leave that for later!

Reply:
Yup, got you!
BTW, I didn't try adding too many funcs to `Multiaddr`. Will be adding them on demand later while implementing other parts